lib/FNAMF/main.c

```diff
@@ -0,0 +1,402 @@
+#include <stdio.h>
+
+#define COBJMACROS
+#include <mfidl.h>
+#include <mfapi.h>
+#include <mfreadwrite.h>
+
+#include "wine/debug.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(fnamf);
+
+struct context
+{
+    IMFSourceReader *reader;
+    DWORD video_stream;
+    DWORD audio_stream;
+
+    IMFMediaBuffer *audio_buffer;
+    IMFMediaBuffer *video_buffer;
+    BOOL eos;
+};
+
+static void set_audio_media_type(struct context *ctx)
+{
+    IMFMediaType *type;
+    HRESULT hr;
+
+    IMFSourceReader_SetStreamSelection(ctx->reader, ctx->audio_stream, FALSE);
+
+    if (SUCCEEDED(hr = MFCreateMediaType(&type)))
+    {
+        hr = IMFMediaType_SetGUID(type, &MF_MT_MAJOR_TYPE, &MFMediaType_Audio);
+
+        if (SUCCEEDED(hr))
+            hr = IMFMediaType_SetGUID(type, &MF_MT_SUBTYPE, &MFAudioFormat_Float);
+
+        if (SUCCEEDED(hr))
+            hr = IMFSourceReader_SetCurrentMediaType(ctx->reader, ctx->audio_stream, NULL, type);
+
+        IMFMediaType_Release(type);
+    }
+
+    IMFSourceReader_SetStreamSelection(ctx->reader, ctx->audio_stream, SUCCEEDED(hr));
+}
+
+static void set_video_media_type(struct context *ctx)
+{
+    IMFMediaType *type;
+    HRESULT hr;
+
+    IMFSourceReader_SetStreamSelection(ctx->reader, ctx->video_stream, FALSE);
+
+    if (SUCCEEDED(hr = MFCreateMediaType(&type)))
+    {
+        hr = IMFMediaType_SetGUID(type, &MF_MT_MAJOR_TYPE, &MFMediaType_Video);
+
+        if (SUCCEEDED(hr))
+            hr = IMFMediaType_SetGUID(type, &MF_MT_SUBTYPE, &MFVideoFormat_I420);
+
+        if (SUCCEEDED(hr))
+            hr = IMFSourceReader_SetCurrentMediaType(ctx->reader, ctx->video_stream, NULL, type);
+
+        IMFMediaType_Release(type);
+    }
+
+    IMFSourceReader_SetStreamSelection(ctx->reader, ctx->video_stream, SUCCEEDED(hr));
+}
+
+void CDECL tf_fopen(const WCHAR *url, struct context **ctx)
+{
+    IMFSourceReader *reader;
+    IMFAttributes *attributes;
+    HRESULT hr;
+
+    WINE_TRACE("url %s, ctx %p.\n", wine_dbgstr_w(url), ctx);
+
+    *ctx = NULL;
+    if (FAILED(hr = MFCreateAttributes(&attributes, 1)))
+        return;
+
+    hr = IMFAttributes_SetUINT32(attributes, &MF_SOURCE_READER_ENABLE_VIDEO_PROCESSING, TRUE);
+    if (SUCCEEDED(hr))
+    {
+        hr = MFCreateSourceReaderFromURL(url, attributes, &reader);
+    }
+    IMFAttributes_Release(attributes);
+
+    if (SUCCEEDED(hr))
+    {
+        if (!(*ctx = calloc(1, sizeof(**ctx))))
+            IMFSourceReader_Release(reader);
+        else
+        {
+            (*ctx)->reader = reader;
+            (*ctx)->video_stream = MF_SOURCE_READER_FIRST_VIDEO_STREAM;
+            (*ctx)->audio_stream = MF_SOURCE_READER_FIRST_AUDIO_STREAM;
+            set_video_media_type(*ctx);
+            set_audio_media_type(*ctx);
+        }
+    }
+
+    WINE_TRACE("ret ctx %p, reader %p, hr %#x.\n", *ctx, reader, hr);
+}
+
+void CDECL tf_close(struct context *ctx)
+{
+    WINE_TRACE("ctx %p.\n", ctx);
+    if (ctx->audio_buffer) IMFMediaBuffer_Release(ctx->audio_buffer);
```

Comment:
Should this also release `ctx->video_buffer`?

Reply:
Ah yeah the video buffer actually should not exist I think.

As I was adding the audio buffer (because sometimes MF decodes more audio than is going to be requested), I thought I would need it for video too, but it turns out that the video player is only requesting one video frame.

This assumes that the decoded video sample comes at one frame at a time, which I think is the case, but I could probably double check.

Reply:
I guess video_buffer is unused.